src/mesh.cpp

```diff
@@ -213,88 +215,105 @@ void Mesh::material_volumes(
     Particle p;
 
     SourceSite site;
-    site.r = {bbox.xmin, 0.0, 0.0};
-    site.u = {1.0, 0.0, 0.0};
     site.E = 1.0;
     site.particle = ParticleType::neutron;
 
-#pragma omp for
-    for (int iy = 0; iy < ny; ++iy) {
-      site.r.y = bbox.ymin + (iy + 0.5) * dy;
-      for (int iz = 0; iz < nz; ++iz) {
-        site.r.z = bbox.zmin + (iz + 0.5) * dz;
-
-        p.from_source(&site);
-
-        // Determine particle's location
-        if (!exhaustive_find_cell(p)) {
-          out_of_model = true;
-          continue;
-        }
-
-        // Set birth cell attribute
-        if (p.cell_born() == C_NONE)
-          p.cell_born() = p.lowest_coord().cell;
-
-        // Initialize last cells from current cell
-        for (int j = 0; j < p.n_coord(); ++j) {
-          p.cell_last(j) = p.coord(j).cell;
-        }
-        p.n_coord_last() = p.n_coord();
-
-        while (true) {
-          // Ray trace from r_start to r_end
-          Position r0 = p.r();
-          double max_distance = bbox.xmax - r0.x;
-
-          // Find the distance to the nearest boundary
-          BoundaryInfo boundary = distance_to_boundary(p);
-
-          // Advance particle forward
-          double distance = std::min(boundary.distance, max_distance);
-          p.move_distance(distance);
+    for (int axis = 0; axis < 3; ++axis) {
+      // Set starting position and direction
+      site.r = {0.0, 0.0, 0.0};
+      site.r[axis] = bbox.min()[axis];
+      site.u = {0.0, 0.0, 0.0};
+      site.u[axis] = 1.0;
+
+      // Determine width of rays and number of rays in other directions
+      int ax1 = (axis + 1) % 3;
+      int ax2 = (axis + 2) % 3;
+      double min1 = bbox.min()[ax1];
+      double min2 = bbox.min()[ax2];
+      double d1 = width[ax1];
+      double d2 = width[ax2];
```

Comment:
So we are assigning a rectangular cross sectional area to the rays, right?

Reply:
Correct, we are assigning a rectangular cross sectional area